keps/sig-api-machinery/0014-garbage-collection.md

```diff
@@ -0,0 +1,555 @@
+[DaemonSets]: https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/
+[CronJobs]: https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/
+[ReplicaSets]: https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/
+
+#### TTL-based Garbage Collection (Events only)
+
+TTL-based garbage collection assigns a TTL (time to live) when creating an
+object (in API server, using [etcd][] TTL) and the API server will delete the
+object once the TTL expires. The TTL can only be assigned at the resource object
+creation time.
+
+Event garbage collection is TTL-based. API server assigns a TTL to a Kubernetes
+Event when creating it, and deletes it when the TTL expires. The same TTL is
+used for all events in the cluster and can be defined via a flag `--event-ttl`
+of [kube-apiserver][].
+
+[etcd]: https://github.com/coreos/etcd
+[kube-apiserver]: https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/
+
+#### Summary
+
+In summary, here are the Kubernetes resources that can be garbage collected.
+
+| Resources                                               | GC Flavor      | GC controller         |
+|---------------------------------------------------------|----------------|-----------------------|
+| Any resource with non-empty `.metadata.ownerReferences` | OwnerReference | Generic GC controller |
+| Terminated Pods                                         | Capacity       | Pod GC controller     |
+| Controller history resources                            | Capacity       | Workload controller   |
+| Events                                                  | TTLs           | API server (etcd TTL) |
+
+### Existing Problems
+
+Even though Kubernetes provides several different flavors of garbage collection,
+there are still some problems that need to be solved. 
+
+#### Non-generic Capacity-based Garbage Collection
+
+Current capacity-based garbage collection has the following issues:
+
+* Not generic. Only supports terminated pods and controller history resources. 
+* Terminated pod garbage collection can only be configured via component flag,
+  and any changes to it requires restarting the component. 
+* No fairness for terminated pod garbage collection. For example, capacity
+  cannot be split equally by namespace. 
+* Terminated pods are deleted in the order of creation time, instead of
+  terminated time. 
+* Capacity has no relation to [resource quotas][]. 
+* Garbage collecting terminated pods without looking at their owner references
+  can potentially break Jobs, because Job depends on the Pods it owns to exist
+  (see [#28486]).
+
+[resource quotas]: https://kubernetes.io/docs/concepts/policy/resource-quotas/
+[#28486]: https://github.com/kubernetes/kubernetes/issues/28486
+
+#### Non-generic TTL-based Garbage Collection
+
+Current TTL-based garbage collection has the following issues:
+
+* Not generic. Only supports Events. 
```

Comment:
that's true for kubernetes resources... other apiservers using the apimachinery storage layer can make use of this today to set variable-length TTLs on objects based on fields in the object

Reply:
Yes. I should clarify that this point is specific to built-in apiserver.